## Importing Libraries

In [1]:
#importing the required libraries
#Data wrangling
import numpy as np
import pandas as pd
pd.set_option("display.max_colwidth",1000)#setting maximum column width

## Reading the datasets

In [13]:
books_df = pd.read_csv('BX-Books.csv', sep=';',error_bad_lines=False, encoding='latin-1')
books_df

C:\Users\ritis\AppData\Local\Temp\ipykernel_30992\3577156960.py:1: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  books_df = pd.read_csv('BX-Books.csv', sep=';',error_bad_lines=False, encoding='latin-1')
Skipping line 6452: expected 8 fields, saw 9
Skipping line 43667: expected 8 fields, saw 10
Skipping line 51751: expected 8 fields, saw 9

Skipping line 92038: expected 8 fields, saw 9
Skipping line 104319: expected 8 fields, saw 9
Skipping line 121768: expected 8 fields, saw 9

Skipping line 144058: expected 8 fields, saw 9
Skipping line 150789: expected 8 fields, saw 9
Skipping line 157128: expected 8 fields, saw 9
Skipping line 180189: expected 8 fields, saw 9
Skipping line 185738: expected 8 fields, saw 9

Skipping line 209388: expected 8 fields, saw 9
Skipping line 220626: expected 8 fields, saw 9
Skipping line 227933: expected 8 fields, saw 11
Skipping line 228957: expected 8 fields, saw

ISBN  \
0       0195153448   
1       0002005018   
2       0060973129   
3       0374157065   
4       0393045218   
...            ...   
271355  0440400988   
271356  0525447644   
271357  006008667X   
271358  0192126040   
271359  0767409752   

                                                                                                                            Book-Title  \
0                                                                                                                  Classical Mythology   
1                                                                                                                         Clara Callan   
2                                                                                                                 Decision in Normandy   
3                                   Flu: The Story of the Great Influenza Pandemic of 1918 and the Search for the Virus That Caused It   
4                                                                                                               The Mummies of Urumchi   
...                                                                                                                                ...   
271355                                                                                                      There's a Bat in Bunk Five   
271356                                                                                                         From One to One Hundred   
271357                                                                   Lily Dale : The True Story of the Town that Talks to the Dead   
271358                                                                                                     Republic (World's Classics)   
271359  A Guided Tour of Rene Descartes' Meditations on First Philosophy with Complete Translations of the Meditations by Ronald Rubin   

                 Book-Author Year-Of-Publication  \
0         Mark P. O. Morford                2002   
1       Richard Bruce Wright                2001   
2               Carlo D'Este                1991   
3           Gina Bari Kolata                1999   
4            E. J. W. Barber                1999   
...                      ...                 ...   
271355        Paula Danziger                1988   
271356            Teri Sloat                1991   
271357      Christine Wicker                2004   
271358                 Plato                1996   
271359   Christopher  Biffle                2000   

                                               Publisher  \
0                                Oxford University Press   
1                                  HarperFlamingo Canada   
2                                        HarperPerennial   
3                                   Farrar Straus Giroux   
4                             W. W. Norton &amp; Company   
...                                                  ...   
271355                   Random House Childrens Pub (Mm)   
271356                                      Dutton Books   
271357                                HarperSanFrancisco   
271358                           Oxford University Press   
271359  McGraw-Hill Humanities/Social Sciences/Languages   

                                                         Image-URL-S  \
0       http://images.amazon.com/images/P/0195153448.01.THUMBZZZ.jpg   
1       http://images.amazon.com/images/P/0002005018.01.THUMBZZZ.jpg   
2       http://images.amazon.com/images/P/0060973129.01.THUMBZZZ.jpg   
3       http://images.amazon.com/images/P/0374157065.01.THUMBZZZ.jpg   
4       http://images.amazon.com/images/P/0393045218.01.THUMBZZZ.jpg   
...                                                              ...   
271355  http://images.amazon.com/images/P/0440400988.01.THUMBZZZ.jpg   
271356  http://images.amazon.com/images/P/0525447644.01.THUMBZZZ.jpg   
271357  http://images.amazon.com/images/P/006008667X.01.THUMBZZZ.jpg   
271358  http://images.amaz

In [10]:
users_df=pd.read_csv("BX-Users.csv", sep=';',error_bad_lines=False, encoding='latin-1')
users_df

C:\Users\ritis\AppData\Local\Temp\ipykernel_30992\703082439.py:1: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  users_df=pd.read_csv("BX-Users.csv", sep=';',error_bad_lines=False, encoding='latin-1')


User-ID                            Location   Age
0             1                  nyc, new york, usa   NaN
1             2           stockton, california, usa  18.0
2             3     moscow, yukon territory, russia   NaN
3             4           porto, v.n.gaia, portugal  17.0
4             5  farnborough, hants, united kingdom   NaN
...         ...                                 ...   ...
278853   278854               portland, oregon, usa   NaN
278854   278855  tacoma, washington, united kingdom  50.0
278855   278856           brampton, ontario, canada   NaN
278856   278857           knoxville, tennessee, usa   NaN
278857   278858                dublin, n/a, ireland   NaN

[278858 rows x 3 columns]

In [11]:
rating_df=pd.read_csv("BX-Book-Ratings.csv", sep=';',error_bad_lines=False, encoding='latin-1')
rating_df

C:\Users\ritis\AppData\Local\Temp\ipykernel_30992\315174063.py:1: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  rating_df=pd.read_csv("BX-Book-Ratings.csv", sep=';',error_bad_lines=False, encoding='latin-1')


User-ID         ISBN  Book-Rating
0         276725   034545104X            0
1         276726   0155061224            5
2         276727   0446520802            0
3         276729   052165615X            3
4         276729   0521795028            6
...          ...          ...          ...
1149775   276704   1563526298            9
1149776   276706   0679447156            0
1149777   276709   0515107662           10
1149778   276721   0590442449           10
1149779   276723  05162443314            8

[1149780 rows x 3 columns]

In [15]:
books_df.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

## Checking for null values and rectifying them.

In [17]:
# Checking  for  null value in book author
books_df[books_df['Book-Author'].isna()]

# Checking null values in publisher
books_df[books_df['Publisher'].isna()]

ISBN       Book-Title      Book-Author Year-Of-Publication  \
128890  193169656X      Tyrant Moon  Elaine Corvidae                2002   
129037  1931696993  Finders Keepers  Linnea Sinclair                2001   

       Publisher  \
128890       NaN   
129037       NaN   

                                                         Image-URL-S  \
128890  http://images.amazon.com/images/P/193169656X.01.THUMBZZZ.jpg   
129037  http://images.amazon.com/images/P/1931696993.01.THUMBZZZ.jpg   

                                                         Image-URL-M  \
128890  http://images.amazon.com/images/P/193169656X.01.MZZZZZZZ.jpg   
129037  http://images.amazon.com/images/P/1931696993.01.MZZZZZZZ.jpg   

                                                         Image-URL-L  
128890  http://images.amazon.com/images/P/193169656X.01.LZZZZZZZ.jpg  
129037  http://images.amazon.com/images/P/1931696993.01.LZZZZZZZ.jpg

In [18]:
#Filling the null value in Book-Author
books_df.loc[187689, 'Book-Author'] = 'Larissa Anne Downes'

#Replacing NaNs with correct  values in Publisher
books_df.loc[128890, 'Publisher'] = 'Mundania Press LLC'
books_df.loc[129037, 'Publisher'] = 'Bantam'

In [19]:
books_df['Year-Of-Publication'].unique()

array([2002, 2001, 1991, 1999, 2000, 1993, 1996, 1988, 2004, 1998, 1994,
       2003, 1997, 1983, 1979, 1995, 1982, 1985, 1992, 1986, 1978, 1980,
       1952, 1987, 1990, 1981, 1989, 1984, 0, 1968, 1961, 1958, 1974,
       1976, 1971, 1977, 1975, 1965, 1941, 1970, 1962, 1973, 1972, 1960,
       1966, 1920, 1956, 1959, 1953, 1951, 1942, 1963, 1964, 1969, 1954,
       1950, 1967, 2005, 1957, 1940, 1937, 1955, 1946, 1936, 1930, 2011,
       1925, 1948, 1943, 1947, 1945, 1923, 2020, 1939, 1926, 1938, 2030,
       1911, 1904, 1949, 1932, 1928, 1929, 1927, 1931, 1914, 2050, 1934,
       1910, 1933, 1902, 1924, 1921, 1900, 2038, 2026, 1944, 1917, 1901,
       2010, 1908, 1906, 1935, 1806, 2021, '2000', '1995', '1999', '2004',
       '2003', '1990', '1994', '1986', '1989', '2002', '1981', '1993',
       '1983', '1982', '1976', '1991', '1977', '1998', '1992', '1996',
       '0', '1997', '2001', '1974', '1968', '1987', '1984', '1988',
       '1963', '1956', '1970', '1985', '1978', '1973', '1980'

Some incorrect names have been put in this column. This needs to be corrected.


In [21]:
books_df[books_df['Year-Of-Publication'] == 'DK Publishing Inc']

ISBN  \
209538  078946697X   
221678  0789466953   

                                                                                                         Book-Title  \
209538         DK Readers: Creating the X-Men, How It All Began (Level 4: Proficient Readers)\";Michael Teitelbaum"   
221678  DK Readers: Creating the X-Men, How Comic Books Come to Life (Level 4: Proficient Readers)\";James Buckley"   

       Book-Author Year-Of-Publication  \
209538        2000   DK Publishing Inc   
221678        2000   DK Publishing Inc   

                                                           Publisher  \
209538  http://images.amazon.com/images/P/078946697X.01.THUMBZZZ.jpg   
221678  http://images.amazon.com/images/P/0789466953.01.THUMBZZZ.jpg   

                                                         Image-URL-S  \
209538  http://images.amazon.com/images/P/078946697X.01.MZZZZZZZ.jpg   
221678  http://images.amazon.com/images/P/0789466953.01.MZZZZZZZ.jpg   

                                                         Image-URL-M  \
209538  http://images.amazon.com/images/P/078946697X.01.LZZZZZZZ.jpg   
221678  http://images.amazon.com/images/P/0789466953.01.LZZZZZZZ.jpg   

       Image-URL-L  
209538         NaN  
221678         NaN

Correction needed in the Book author and Year of Publication

In [ ]:
# on searching for these  books we came to know about its authors
#ISBN '078946697X'
books_df.loc[books_df.ISBN == '078946697X','Year-Of-Publication'] = 2000
books_df.loc[books_df.ISBN == '078946697X','Book-Author'] = "Michael Teitelbaum"
books_df.loc[books_df.ISBN == '078946697X','Publisher'] = "DK Publishing Inc"
books_df.loc[books_df.ISBN == '078946697X','Book-Title'] = "DK Readers: Creating the X-Men, How It All Began (Level 4: Proficient Readers)"

#ISBN '0789466953'
books_df.loc[books_df.ISBN == '0789466953','Year-Of-Publication'] = 2000
books_df.loc[books_df.ISBN == '0789466953','Book-Author'] = "James Buckley"
books_df.loc[books_df.ISBN == '0789466953','Publisher'] = "DK Publishing Inc"
books_df.loc[books_df.ISBN == '0789466953','Book-Title'] = "DK Readers: Creating the X-Men, How Comic Books Come to Life"

In [24]:
#checking the rows having 'Gallimard' as yearOfPublication
books_df.loc[books_df['Year-Of-Publication'] == 'Gallimard',:]

ISBN  \
220731  2070426769   

                                                                      Book-Title  \
220731  Peuple du ciel, suivi de 'Les Bergers\";Jean-Marie Gustave Le Cl�©zio"   

       Book-Author Year-Of-Publication  \
220731        2003           Gallimard   

                                                           Publisher  \
220731  http://images.amazon.com/images/P/2070426769.01.THUMBZZZ.jpg   

                                                         Image-URL-S  \
220731  http://images.amazon.com/images/P/2070426769.01.MZZZZZZZ.jpg   

                                                         Image-URL-M  \
220731  http://images.amazon.com/images/P/2070426769.01.LZZZZZZZ.jpg   

       Image-URL-L  
220731         NaN

Correction needed in Book Author and Year of Publication

In [25]:
books_df.loc[books_df.ISBN=='2070426769','Year-Of-Publication']=2003
books_df.loc[books_df.ISBN=='2070426769','Book-Author']='Jean-Marie Gustave Le Cl�©zio'
books_df.loc[books_df.ISBN=='2070426769','Publisher']='Gallimard'
books_df.loc[books_df.ISBN=='2070426769','Book-Title']="Peuple du ciel, suivi de 'Les Bergers"

In [26]:
books_df['Year-Of-Publication'] =books_df['Year-Of-Publication'].astype(int)

#correction in books having year of publication < 1800
books_df.loc[books_df.ISBN=='	9643112136','Year-Of-Publication'] = 2010
books_df.loc[books_df.ISBN=='964442011X', 'Year-Of-Publication'] = 1991
     

In [27]:
#Sustituting np.Nan in rows with year=0 or  greater than the current year,2022.
books_df.loc[(books_df['Year-Of-Publication'] > 2022) | (books_df['Year-Of-Publication'] == 0),'Year-Of-Publication'] = np.NAN

# replacing NaN values with median value of Year-Of-Publication
books_df['Year-Of-Publication'].fillna(int(books_df['Year-Of-Publication'].median()), inplace=True)

In [28]:
#Uppercasing the ISBN numbers and Book-Author name
books_df['ISBN']=books_df['ISBN'].str.upper()
books_df['Book-Author']=books_df['Book-Author'].str.upper()

### Checking for Duplicates and rectifying them

In [29]:
books_df[books_df.duplicated()]  

ISBN  \
6635    002542730X   
11925   014062080X   
16298   097089726X   
19266   039592720X   
22339   042512164X   
...            ...   
267020  088404632X   
269297  055215038X   
269552  037376099X   
270285  189481505X   
270483  039304632X   

                                                                      Book-Title  \
6635    Politically Correct Bedtime Stories: Modern Tales for Our Life and Times   
11925                              The Scarlet Letter (Penguin Popular Classics)   
16298                                                                     Jasper   
19266                                                    Interpreter of Maladies   
22339                                                             Capital Crimes   
...                                                                          ...   
267020                            Dianetics: The Modern Science of Mental Health   
269297                                                               The Sleeper   
269552             Patient Nurse  (Man Of The Month) (Harlequin Desire, No 1099)   
270285                                                                       I-O   
270483                                                 The Voyage of the Narwhal   

                Book-Author  Year-Of-Publication                  Publisher  \
6635      JAMES FINN GARNER               1994.0  John Wiley &amp; Sons Inc   
11925   NATHANIEL HAWTHORNE               1994.0          Penguin Books Ltd   
16298        MICHELLE GROCE               2003.0     Novello Festival Press   
19266         JHUMPA LAHIRI               1999.0        Houghton Mifflin Co   
22339      LAWRENCE SANDERS               2000.0   Berkley Publishing Group   
...                     ...                  ...                        ...   
267020       L. RON HUBBARD               1995.0   Bridge Publications (CA)   
269297        GILLIAN WHITE               2003.0                Corgi Books   
269552         DIANA PALMER               1997.0                 Silhouette   
270285          SIMON LOGAN               2002.0                 Prime (OH)   
270483       ANDREA BARRETT               1998.0    W W Norton &amp; Co Inc   

                                                         Image-URL-S  \
6635    http://images.amazon.com/images/P/002542730X.01.THUMBZZZ.jpg   
11925   http://images.amazon.com/images/P/014062080X.01.THUMBZZZ.jpg   
16298   http://images.amazon.com/images/P/097089726X.01.THUMBZZZ.jpg   
19266   http://images.amazon.com/images/P/039592720X.01.THUMBZZZ.jpg   
22339   http://images.amazon.com/images/P/042512164X.01.THUMBZZZ.jpg   
...                                                              ...   
267020  http://images.amazon.com/images/P/088404632X.01.THUMBZZZ.jpg   
269297  http://images.amazon.com/images/P/055215038X.01.THUMBZZZ.jpg   
269552  http://images.amazon.com/images/P/037376099X.01.THUMBZZZ.jpg   
270285  http://images.amazon.com/images/P/189481505X.01.THUMBZZZ.jpg   
270483  http://images.amazon.com/images/P/039304632X.01.THUMBZZZ.jpg   

                                                         Image-URL-M  \
6635    http://images.amazon.com/images/P/002542730X.01.MZZZZZZZ.jpg   
11925   http://images.amazon.com/images/P/014062080X.01.MZZZZZZZ.jpg   
16298   http://images.amazon.com/images/P/097089726X.01.MZZZZZZZ.jpg   
19266   http://images.amazon.com/images/P/039592720X.01.MZZZZZZZ.jpg   
22339   http://images.amazon.com/images/P/042512164X.01.MZZZZZZZ.jpg   
...                                                              ...   
267020  http://images.amazon.com/images/P/088404632X.01.MZZZZZZZ.jpg   
269297  http://images.amazon.com/images/P/055215038X.01.MZZZZZZZ.jpg   
269552  http://images.amazon.com/images/P/037376099X.01.MZZZZZZZ.jpg   
270285  http://images.amazon.com/images/P/189481505X.01.MZZZZZZZ.jpg   
270483  http://images.amazon.com/images/P/039304632X.01.MZZZZZZZ.jpg   

                                                         Imag

In [30]:
#Dropping the rows with the entire column values are duplicated
books_df.drop_duplicates(keep="first",inplace=True)
books_df.reset_index(drop=True,inplace=True) 

In [31]:
#Dropping the image urls
books_df.drop(columns=['Image-URL-S','Image-URL-M','Image-URL-L'],inplace=True)
     

In [32]:
books_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271047 entries, 0 to 271046
Data columns (total 5 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   ISBN                 271047 non-null  object 
 1   Book-Title           271047 non-null  object 
 2   Book-Author          271047 non-null  object 
 3   Year-Of-Publication  271047 non-null  float64
 4   Publisher            271047 non-null  object 
dtypes: float64(1), object(4)
memory usage: 10.3+ MB


All the null values are rectified

## Data Preprocessing in User data

In [33]:
users_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   User-ID   278858 non-null  int64  
 1   Location  278858 non-null  object 
 2   Age       168096 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB


Age contains null values which need to be rectified

In [34]:
users_df.loc[(users_df['Age']>95)|(users_df['Age']<4),'Age']=np.nan

In [35]:
users_df['Age'].describe()

count    166857.000000
mean         34.734222
std          13.621521
min           4.000000
25%          24.000000
50%          32.000000
75%          44.000000
max          95.000000
Name: Age, dtype: float64

In [36]:
def age_group(age):
  ''' 
  defines the age group of users
  '''
  if age<13:
    x='Children'
  elif age>=13 and age<18:
    x='Teens'
  elif age>=18 and age<36:
    x='Youth'
  elif age>=36 and age<56:
    x='Middle aged adults'
  else:
    x='Elderly'
  return x
     


In [37]:
users_df['Age_group']=users_df['Age'].apply(lambda x: age_group(x))

In [38]:
#Dropping the column 'Location'
users_df.drop('Location',axis=1,inplace=True)

## Data Preprocessing on Ratings Data

In [39]:
rating_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   User-ID      1149780 non-null  int64 
 1   ISBN         1149780 non-null  object
 2   Book-Rating  1149780 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 26.3+ MB


no null values in this dataset

In [40]:
# making all the ISBN no. uppercase
rating_df['ISBN'].apply(lambda x: x.upper())

0           034545104X
1           0155061224
2           0446520802
3           052165615X
4           0521795028
              ...     
1149775     1563526298
1149776     0679447156
1149777     0515107662
1149778     0590442449
1149779    05162443314
Name: ISBN, Length: 1149780, dtype: object

In [41]:
# lets see if all the books in rating_df are also in books_df
rating_df_new = rating_df[rating_df['ISBN'].isin(books_df['ISBN'])]

In [42]:
rating_df_new['Book-Rating'].value_counts().reset_index()

index  Book-Rating
0       0       646974
1       8        91739
2      10        71194
3       7        66356
4       9        60754
5       5        45343
6       6        31673
7       4         7613
8       3         5117
9       2         2374
10      1         1480

We can see for explicit ratings that rating 8 is received by most number of books followed by rating 10. There are 646974 implicitely rated books.

In [43]:
explicit_rating = rating_df_new[rating_df_new['Book-Rating'] != 0]
implicit_rating = rating_df_new[rating_df_new['Book-Rating'] == 0]

## Merging Datasets

In [44]:
# for the rating dataset, we are only taking the explicit rating dataset
df = pd.merge(books_df, explicit_rating, on='ISBN', how='inner')
df = pd.merge(df, users_df, on='User-ID', how='inner')

In [45]:
#displaying the top 3 rows of df
df.head(3)

ISBN                               Book-Title           Book-Author  \
0  0002005018                             Clara Callan  RICHARD BRUCE WRIGHT   
1  074322678X  Where You'll Find Me: And Other Stories           ANN BEATTIE   
2  0887841740                       The Middle Stories           SHEILA HETI   

   Year-Of-Publication              Publisher  User-ID  Book-Rating  Age  \
0               2001.0  HarperFlamingo Canada        8            5  NaN   
1               2002.0               Scribner        8            5  NaN   
2               2004.0  House of Anansi Press        8            5  NaN   

  Age_group  
0   Elderly  
1   Elderly  
2   Elderly

## Using Collaborative Filtering

### - Model based approach using SVD

In [46]:
from sklearn.model_selection import train_test_split
from scipy.sparse.linalg import svds
import random
import math
     

In [47]:
#Filtering books with more than 5 reviews

print('The number of books that are explicitely rated are',explicit_rating['ISBN'].nunique())
ratings_count_df=explicit_rating.groupby("ISBN")['User-ID'].count().to_frame('No-of-rated-users').reset_index()
selected_books =list(ratings_count_df.loc[ratings_count_df['No-of-rated-users']>5,'ISBN'].unique())
print('Number of  books rated by atleast 5 users:',len(selected_books))
filter_df=explicit_rating.loc[explicit_rating['ISBN'].isin(selected_books)]

The number of books that are explicitely rated are 149655
Number of  books rated by atleast 5 users: 10731


In [48]:
#keeping books with selected users
print('The number of users who have explicitely rated books are',explicit_rating['User-ID'].nunique())

#keeps Users who have rated more than five books
books_count_df=filter_df.groupby("User-ID")['ISBN'].count().to_frame('No-of-books-rated').reset_index()
selected_users = list(books_count_df.loc[books_count_df['No-of-books-rated']>5,'User-ID'].unique())
print('Number of  users who have rated atleast 5 books are :',len(selected_users))

#dataframe with filtered number of interactions
filter_df=filter_df.loc[filter_df['User-ID'].isin(selected_users)]
print('The shape of data fame with filtered number of interactions : ',filter_df.shape)
     

The number of users who have explicitely rated books are 68080
Number of  users who have rated atleast 5 books are : 5911
The shape of data fame with filtered number of interactions :  (114147, 3)


In [54]:
complete_df = filter_df.copy()


In [55]:
train_df, test_df = train_test_split(complete_df,
                                   stratify=complete_df['User-ID'], 
                                   test_size=0.20,
                                   random_state=0)



## Creating a sparse pivot table

In [56]:
users_books_pivot_matrix_df = train_df.pivot(index='User-ID', 
                                                          columns='ISBN', 
                                                          values='Book-Rating').fillna(0)

users_books_pivot_matrix_df.head()

ISBN     0002005018  0002251760  0002259834  0002558122  0006480764  \
User-ID                                                               
114             0.0         0.0         0.0         0.0         0.0   
243             0.0         0.0         0.0         0.0         0.0   
244             0.0         0.0         0.0         0.0         0.0   
254             0.0         0.0         0.0         0.0         0.0   
388             0.0         0.0         0.0         0.0         0.0   

ISBN     000648302X  000649840X  0006546684  0006547834  0006550576  ...  \
User-ID                                                              ...   
114             0.0         0.0         0.0         0.0         0.0  ...   
243             0.0         0.0         0.0         0.0         0.0  ...   
244             0.0         0.0         0.0         0.0         0.0  ...   
254             0.0         0.0         0.0         0.0         0.0  ...   
388             0.0         0.0         0.0         0.0         0.0  ...   

ISBN     881787017X  8838910170  8845205118  8845247414  8845915611  \
User-ID                                                               
114             0.0         0.0         0.0         0.0         0.0   
243             0.0         0.0         0.0         0.0         0.0   
244             0.0         0.0         0.0         0.0         0.0   
254             0.0         0.0         0.0         0.0         0.0   
388             0.0         0.0         0.0         0.0         0.0   

ISBN     9074336329  950491036X  9681500555  968150108X  9726101794  
User-ID                                                              
114             0.0         0.0         0.0         0.0         0.0  
243             0.0         0.0         0.0         0.0         0.0  
244             0.0         0.0         0.0         0.0         0.0  
254             0.0         0.0         0.0         0.0         0.0  
388             0.0         0.0         0.0         0.0         0.0  

[5 rows x 10617 columns]

## Creating a factorized matrix

In [57]:
#Creating a matrix with the values of users_books_pivot_matrix_df
original_ratings_matrix = users_books_pivot_matrix_df.values

#Storing the User-IDs in a list
user_ids = list(users_books_pivot_matrix_df.index)

# The number of factors to factor the user-item matrix.
NUMBER_OF_FACTORS_MF = 20

#Performs matrix factorization of the original user item matrix
U, sigma, Vt = svds(original_ratings_matrix, k = NUMBER_OF_FACTORS_MF)
     

#converting sigma to a diagonal matrix
sigma = np.diag(sigma)
     

After the factorization the resulting matrix is not sparse any more. It has generated rating predictions for books with which users have not yet interacted (and therefore not rated), which will be used to recommend relevant books to the user.

In [58]:
#Rating matric reconstructed using the matrices obtained after factorizing
predicted_ratings_matrix = np.dot(np.dot(U, sigma), Vt) 
predicted_ratings_matrix 

array([[-0.0112932 ,  0.01781487,  0.00036885, ...,  0.00285667,
         0.0062242 , -0.00086892],
       [ 0.0024454 ,  0.03261562,  0.00080351, ...,  0.00698411,
        -0.00770924,  0.00166759],
       [ 0.0057473 ,  0.00875379,  0.0009157 , ...,  0.00026397,
        -0.00093596,  0.00056328],
       ...,
       [-0.00670329,  0.0121104 ,  0.00184951, ..., -0.00477621,
         0.00093827,  0.000227  ],
       [-0.0037192 , -0.01728346, -0.00106902, ..., -0.00182796,
        -0.00637287,  0.00884846],
       [ 0.00777082, -0.00139017,  0.00027011, ..., -0.01105348,
        -0.00021274,  0.0014974 ]])

In [60]:
#Converting the reconstructed matrix back to a Pandas dataframe
predicted_ratings_df = pd.DataFrame(predicted_ratings_matrix , columns = users_books_pivot_matrix_df.columns, index=user_ids).transpose()

     

## Building Recommender Model

In [61]:
class CFRecommender:
    #Storing model name
    MODEL_NAME = 'Collaborative Filtering'
    
    def __init__(self, cf_predictions_df, items_df=None):
        #Creating attributes
        self.cf_predictions_df = cf_predictions_df
        self.items_df = items_df
        
    def get_model_name(self):
        '''This will return model name'''
        return self.MODEL_NAME
        
    def recommend_items(self, user_id, items_to_ignore=[], topn=10, verbose=False):
        # Get and sort the user's predictions
        sorted_user_predictions = self.cf_predictions_df[user_id].sort_values(ascending=False).reset_index().rename(columns={user_id: 'Book-Rating'})

        # Recommend the highest predicted rating content that the user hasn't seen yet.
        recommendations_df = sorted_user_predictions[~sorted_user_predictions['ISBN'].isin(items_to_ignore)].sort_values('Book-Rating', ascending = False).head(topn)

        if verbose:
            #runs only if verbose=True
            if self.items_df is None:
                raise Exception('"items_df" is required in verbose mode')
            #Merging 
            recommendations_df = recommendations_df.merge(self.items_df, how = 'left', 
                                                          left_on = 'ISBN', 
                                                          right_on = 'ISBN')[["ISBN",'Book-Title',	'Book-Author','Publisher']]

        return recommendations_df

#Creating object of the class   
cf_recommender_model = CFRecommender(predicted_ratings_df, books_df)
     

In [62]:
def get_items_interacted(person_id, interactions_df):
    interacted_items = interactions_df.loc[person_id]['ISBN']
    #Repetation is avoided by taking set
    return set(interacted_items if type(interacted_items) == pd.Series else [interacted_items])

In [63]:
#Indexing by personId to speed up the searches during evaluation
full_indexed_df =complete_df.set_index('User-ID')
train_indexed_df = train_df.set_index('User-ID')
test_indexed_df = test_df.set_index('User-ID')
     

### Predicting the books for user having user_id = 254

Firsly finding the preferences of this user (books user has already rated).

In [64]:
print(f'These are  books that the user ID {user_ids[3]} has already rated \n')
books_df.loc[books_df['ISBN'].isin(list(get_items_interacted(user_ids[3],train_indexed_df)))]['Book-Title']

These are  books that the user ID 254 has already rated 



274                            The Golden Compass (His Dark Materials, Book 1)
282                                                        Making Minty Malone
285                                                                Animal Farm
356                                                    The Secret Life of Bees
485                                      She's Come Undone (Oprah's Book Club)
1175                                                             American Gods
2785                                       The Hobbit: or There and Back Again
2809                            Harry Potter and the Sorcerer's Stone (Book 1)
2969                                                 The Bonesetter's Daughter
3459                          Harry Potter and the Chamber of Secrets (Book 2)
3839                         Harry Potter and the Prisoner of Azkaban (Book 3)
4241                                                    American Gods: A Novel
5431                              Harry Potter and t

In [65]:
#Recommending books
print(f'Recommending books for User ID: {user_ids[3]} ')
cf_recommender_model.recommend_items(user_ids[3],items_to_ignore= get_items_interacted(user_ids[3],train_indexed_df),verbose=True)
     

Recommending books for User ID: 254 


ISBN  \
0  043935806X   
1  059035342X   
2  0439139600   
3  0345339681   
4  0385504209   
5  0345313860   
6  0345339738   
7  0385484518   
8  0345337662   
9  0316769487   

                                                                  Book-Title  \
0                         Harry Potter and the Order of the Phoenix (Book 5)   
1           Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))   
2                               Harry Potter and the Goblet of Fire (Book 4)   
3               The Hobbit : The Enchanting Prelude to The Lord of the Rings   
4                                                          The Da Vinci Code   
5                           The Vampire Lestat (Vampire Chronicles, Book II)   
6                     The Return of the King (The Lord of the Rings, Part 3)   
7  Tuesdays with Morrie: An Old Man, a Young Man, and Life's Greatest Lesson   
8                                                 Interview with the Vampire   
9                                                     The Catcher in the Rye   

      Book-Author               Publisher  
0   J. K. ROWLING              Scholastic  
1   J. K. ROWLING  Arthur A. Levine Books  
2   J. K. ROWLING   Scholastic Paperbacks  
3  J.R.R. TOLKIEN                 Del Rey  
4       DAN BROWN               Doubleday  
5       ANNE RICE        Ballantine Books  
6  J.R.R. TOLKIEN                 Del Rey  
7     MITCH ALBOM               Doubleday  
8       ANNE RICE        Ballantine Books  
9   J.D. SALINGER           Little, Brown

### Similarly recommending books for more people

In [67]:
print(f'Recommending books for User ID: {user_ids[5]} ')
cf_recommender_model.recommend_items(user_ids[5],items_to_ignore= get_items_interacted(user_ids[5],train_indexed_df),verbose=True)
     

Recommending books for User ID: 446 


ISBN  \
0  0312195516   
1  0060928336   
2  0452282152   
3  0142001740   
4  067976402X   
5  0446672211   
6  0316601950   
7  0345337662   
8  0385484518   
9  0060930535   

                                                                  Book-Title  \
0                                        The Red Tent (Bestselling Backlist)   
1                            Divine Secrets of the Ya-Ya Sisterhood: A Novel   
2                                                  Girl with a Pearl Earring   
3                                                    The Secret Life of Bees   
4                                                     Snow Falling on Cedars   
5                         Where the Heart Is (Oprah's Book Club (Paperback))   
6                                                 The Pilot's Wife : A Novel   
7                                                 Interview with the Vampire   
8  Tuesdays with Morrie: An Old Man, a Young Man, and Life's Greatest Lesson   
9                                              The Poisonwood Bible: A Novel   

          Book-Author          Publisher  
0       ANITA DIAMANT        Picador USA  
1       REBECCA WELLS          Perennial  
2     TRACY CHEVALIER        Plume Books  
3       SUE MONK KIDD      Penguin Books  
4      DAVID GUTERSON  Vintage Books USA  
5        BILLIE LETTS       Warner Books  
6        ANITA SHREVE     Back Bay Books  
7           ANNE RICE   Ballantine Books  
8         MITCH ALBOM          Doubleday  
9  BARBARA KINGSOLVER          Perennial

In [68]:
print(f'Recommending books for User ID: {user_ids[7]} ')
cf_recommender_model.recommend_items(user_ids[7],items_to_ignore= get_items_interacted(user_ids[7],train_indexed_df),verbose=True)
     

Recommending books for User ID: 505 


ISBN                                                   Book-Title  \
0  0345337662                                   Interview with the Vampire   
1  034538475X  The Tale of the Body Thief (Vampire Chronicles (Paperback))   
2  0345351525     The Queen of the Damned (Vampire Chronicles (Paperback))   
3  0345384466             The Witching Hour (Lives of the Mayfair Witches)   
4  0451163524            The Drawing of the Three (The Dark Tower, Book 2)   
5  0452282152                                    Girl with a Pearl Earring   
6  0345409671                 Memnoch the Devil (Vampire Chronicles, No 5)   
7  0451160525                      The Gunslinger (The Dark Tower, Book 1)   
8  0670839531                   Needful Things: The Last Castle Rock Story   
9  0446310786                                        To Kill a Mockingbird   

       Book-Author                   Publisher  
0        ANNE RICE            Ballantine Books  
1        ANNE RICE            Ballantine Books  
2        ANNE RICE            Ballantine Books  
3        ANNE RICE            Ballantine Books  
4     STEPHEN KING                 Signet Book  
5  TRACY CHEVALIER                 Plume Books  
6        ANNE RICE            Ballantine Books  
7     STEPHEN KING        New American Library  
8     STEPHEN KING                Viking Books  
9       HARPER LEE  Little Brown &amp; Company

In [69]:
print(f'Recommending books for User ID: {user_ids[10]} ')
cf_recommender_model.recommend_items(user_ids[10],items_to_ignore= get_items_interacted(user_ids[10],train_indexed_df),verbose=True)
     

Recommending books for User ID: 643 


ISBN                                         Book-Title  \
0  0446310786                              To Kill a Mockingbird   
1  0345370775                                      Jurassic Park   
2  0439064864   Harry Potter and the Chamber of Secrets (Book 2)   
3  0439136350  Harry Potter and the Prisoner of Azkaban (Book 3)   
4  0312924585                               Silence of the Lambs   
5  0439139597       Harry Potter and the Goblet of Fire (Book 4)   
6  0451526341                                        Animal Farm   
7  0590353403     Harry Potter and the Sorcerer's Stone (Book 1)   
8  0440211727                                     A Time to Kill   
9  044021145X                                           The Firm   

        Book-Author                     Publisher  
0        HARPER LEE    Little Brown &amp; Company  
1  MICHAEL CRICHTON              Ballantine Books  
2     J. K. ROWLING                    Scholastic  
3     J. K. ROWLING                    Scholastic  
4     THOMAS HARRIS            St. Martin's Press  
5     J. K. ROWLING                    Scholastic  
6     GEORGE ORWELL                        Signet  
7     J. K. ROWLING                    Scholastic  
8      JOHN GRISHAM                          Dell  
9      JOHN GRISHAM  Bantam Dell Publishing Group